llvm_ir_dataset_utils/compile_time_analysis_tools/dataset_download/write_data_files.py

```diff
@@ -0,0 +1,125 @@
+  slightly speed up load times. Then, a list of dictionaries is made,
+  where each dict contains the starting and ending index for each IR
+  file type based on language. For example, if the entire dataset
+  consisted of C and C++ IR modules, then the dictionary for C would
+  note language='c', starting_index=0, ending_index=(C++ starting
+  index). The +1 for the ending_index allows for direct use in range(a,
+  b) syntax, but is NOT suitable for right inclusive syntax. The entries
+  of each dictionary are then written to a CSV file name 'indices.csv'
+  for further use by other scripts. The contents of each dictionary are
+  used to provide information to the n number of threads, where n is the
+  number of languages in the dataset. Each thread when started calls the
+  create_tar() sub-function. The sub-function uses a generator
+  expression to access the bitcode files from a subset (taken by
+  language) of the original Dataset object. This allows for low memory
+  usage while performing in-memory writing of each bitcode file to a tar
+  archive which is named according to the given language (i.e.,
+  c_bc_files.tar).
+
+  Args:
+    storage: Storage location for the tar archives
+  """
+
+  def create_tar(dataset_subset, start_index: int, dir_name: str,
+                 language: str):
+    with tarfile.open(dir_name + '/' + language + '_bc_files.tar', 'a:') as tar:
+      for x in enumerate((dataset_subset[i]["content"]
+                          for i in range(0, dataset_subset.num_rows))):
+        tarinfo = tarfile.TarInfo(name=f'bc_files/file{x[0]+1+start_index}.bc')
+        file_obj = BytesIO(x[1])
+        tarinfo.size = file_obj.getbuffer().nbytes
+        tarinfo.mtime = time()
+        tar.addfile(tarinfo, fileobj=file_obj)
+
+  with parallel.parallel_backend('spark'):
+    dataset = load_dataset('llvm-ml/ComPile', split='train', num_proc=2)
+
+  lang_list: [str] = dataset["language"]
+  langs = dataset.unique("language")
+  file_indices: [dict] = []
+
+  for i in range(0, len(langs)):
+    start_index = lang_list.index(langs[i])
+    if (i + 1 != len(langs)):
+      end_index = lang_list.index(langs[i + 1])
+    else:
+      end_index = len(lang_list)
+    file_indices.append({
+        "language": langs[i],
+        "start_index": start_index,
+        "end_index": end_index
+    })
+    with open('indices.csv', mode='w', newline='') as file:
+      writer = csv.DictWriter(
+          file,
+          fieldnames=["language", "start_index", "end_index"],
+          dialect='unix',
+          quoting=csv.QUOTE_NONE)
+      writer.writeheader()
+      writer.writerows(file_indices)
+
+  threads = []
+  for i in range(0, len(file_indices)):
+    start_index = file_indices[i]["start_index"]
+    end_index = file_indices[i]["end_index"]
+    dir_name = f'{storage}/{file_indices[i]["language"]}'
+    makedirs(dir_name, exist_ok=True)
+    thread = threading.Thread(
```

Comment:
It would probably be more natural to use a `ThreadPoolExecutor` here, submit jobs, and get back futures?